### PR TITLE
Fix leak of signal with deallocated generator observer

### PR DESCRIPTION
**Pull request by Vadim-Yelagin**
Fix leak of signal with deallocated generator observer by capturing self weakly in disposable from `Signal.observe`.
Fixes https://github.com/ReactiveCocoa/ReactiveCocoa/issues/2586 , test cases included, all pass.

### CODE REVIEW

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 1]
```diff
@@ -453,6 +473,33 @@ class SignalSpec: QuickSpec {
...
 				expect(lastValue).to(equal("2"))
 			}
+
+			it("should not keep resulting signal alive indefinitely") {
+				weak var signal: Signal<AnyObject, NoError>? = Signal.never.map { $0 }
+
+				expect(signal).to(beNil())
+			}
+
+			it("should not keep resulting signal alive indefinitely even if it has an observer") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Sorry but this changes [the semantics of `Signal`](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/23dab6c3e69a64fb4d027e29ff90ac6896a7bac8/ReactiveCocoa/Swift/Signal.swift#L15-L16).`

Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: @NachoSoto What that means is that the signal is kept alive by its generator observer. If it has no generator observer, then it cannot terminated anymore because there is no way to have new events. In that case there are two possibilities: it either leaks or can deallocate. This PR does exactly the latter. That comment doesn't state anything about the signal being observed or that a disposable from an observation is being retained somewhere (this is what kept it from deallocation). In fact, this test case succeeds even without this PR because it does not retain the disposable. It's the next case that failed.
Here are some relevant lines from the Signal contract (https://github.com/ReactiveCocoa/ReactiveCocoa/blob/master/Documentation/DesignGuidelines.md#the-signalproducer-contract):
> The work associated with a Signal does not start or stop when observers are added or removed

> A signal is retained until the underlying observer is released

(underlying observer is generator observer, not observers of the signal)

> Note that it is possible to release a signal before a terminating event has been sent upon it. This should usually be avoided, as it can result in resource leaks, but is sometimes useful to disable termination.

Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: > In fact, this test case succeeds

Sorry, that's "should not keep signal alive indefinitely even if it has an observer" that passes, the one without map (which should not make any difference). Anyway, if signal is to be alive until it terminates (even if can never terminate) then "should not keep signal alive indefinitely" (that had been there for a long time) should also fail. The only way for it not to fail is to have a retain cycle inside `Signal.never` itself, which is ridiculous. Btw deallocation of a signal here does not affect (stop) any underlying work (in case it's still being performed, but lost its reference to the generator observer and is not going to send events anymore) because it does not call the generator disposable (which could stop the work) as it would if signal terminated.

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 2]
```diff
@@ -453,6 +473,33 @@ class SignalSpec: QuickSpec {
...
+				weak var signal: Signal<AnyObject, NoError>? = {
+					let signal: Signal<AnyObject, NoError> = Signal.never.map { $0 }
+					signal.observe(Observer())
+					return signal
+				}()
+				expect(signal).to(beNil())
+			}
+
+			it("should not keep resulting signal alive indefinitely even if it has an observer with retained disposable") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: This test is repeated, right?

**Comment by NachoSoto** [comment]
@neilpa does this look good to you guys?

File: ReactiveCocoaTests/Swift/SignalSpec.swift (hunk 2 above)
Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: This test is different because unlike the previous one it has a disposable from `observe` retained.
Unlike such tests without `map` they both failed before the fix though, because `map` already retains the disposable.
This is how tests were prior to the fix:
init:
✅should not keep signal alive indefinitely
✅should not keep signal alive indefinitely even if it has an observer
❌should not keep signal alive indefinitely even if it has an observer with retained disposable
map:
❌should not keep resulting signal alive indefinitely
❌should not keep resulting signal alive indefinitely even if it has an observer
❌should not keep resulting signal alive indefinitely even if it has an observer with retained disposable
I've added all variations for `map` for completeness, but if you think it's excessive, I can leave only one.

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 3]
```diff
@@ -36,6 +36,26 @@ class SignalSpec: QuickSpec {
 				expect(signal).to(beNil())
 			}
 
+			it("should not keep signal alive indefinitely even if it has an observer") {
+				weak var signal: Signal<AnyObject, NoError>? = {
+					let signal: Signal<AnyObject, NoError> = Signal.never
```
Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Can we make this explicit (e.g. `Signal.init { _ in }`) and tweak the test description to "should deallocate the signal if it never sends events"

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 4]
```diff
@@ -36,6 +36,26 @@ class SignalSpec: QuickSpec {
...
+					return signal
+				}()
+				expect(signal).to(beNil())
+			}
+
+			it("should not keep signal alive indefinitely even if it has an observer with retained disposable") {
```
Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Can we rephrase this as "should deallocate the signal ..." instead of "should not keep the signal alive indefinitely ...". Same for the rest of the tests below.

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 5]
```diff
@@ -36,6 +36,26 @@ class SignalSpec: QuickSpec {
...
+				var disposable: Disposable? = nil
+				weak var signal: Signal<AnyObject, NoError>? = {
+					let signal: Signal<AnyObject, NoError> = Signal.never
```
Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Can we be explicit here with `Signal.init { _ in }` and in the other tests.

File: ReactiveCocoaTests/Swift/SignalSpec.swift [hunk 6]
```diff
@@ -453,6 +473,33 @@ class SignalSpec: QuickSpec {
 				observer.sendNext(1)
 				expect(lastValue).to(equal("2"))
 			}
+
+			it("should not keep resulting signal alive indefinitely") {
+				weak var signal: Signal<AnyObject, NoError>? = Signal.never.map { $0 }
```
Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Can we split these out into their own set of tests that don't rely on map and instead have a custom transformer

Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: Basically, either these should exist for every operator (which is tedious and error prone) or we just test it once with a custom test operator

File: ReactiveCocoaTests/Swift/SignalSpec.swift (hunk 2 above)
Comment (neilpa) [inline ReactiveCocoaTests/Swift/SignalSpec.swift]: This set seems reasonable. Although instead of using `map` as I mention above these can be organized as a set of deallocation tests both with and w/out a transform.

**Comment by neilpa** [comment]
Some minor formatting and organizational comments, but overall this seems like a good fix to me.

Additionally, we should also just remove `Signal.never`. I see no reason to have that as part of the API.

**Comment by NachoSoto** [comment]
I'll take another look at this after the changes @neilpa suggested.

> Additionally, we should also just remove Signal.never. I see no reason to have that as part of the API.

:+1:

**Comment by Vadim-Yelagin** [comment]
@neilpa do you want these tests in a separate source file (e.g. SignalLifetimeSpec.swift), in a separate `describe("??")` in SignalSpec.swift, or something else?

**Comment by Vadim-Yelagin** [comment]
@NachoSoto @neilpa I've made the changes. Not sure it's exactly what you meant though : )

File: ReactiveCocoa/Swift/Signal.swift [hunk 7]
```diff
@@ -128,8 +128,8 @@ public final class Signal<Value, Error: ErrorType> {
 		}
 
 		if let token = token {
-			return ActionDisposable {
-				self.atomicObservers.modify { observers in
+			return ActionDisposable { [weak self] in
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/Signal.swift]: Curious: do tests pass with this as `unowned`? I'd rather do that since it defines the relationship more correctly.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+//  Created by Vadim Yelagin on 2015-12-13.
+//  Copyright (c) 2015 GitHub. All rights reserved.
+//
+
+import Result
+import Nimble
+import Quick
+import ReactiveCocoa
+
+class SignalLifetimeSpec: QuickSpec {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: I'm :+1: for separating these. Good idea.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+					let signal: Signal<AnyObject, NoError> = Signal { _ in nil }
+					disposable = signal.observe(Observer())
+					return signal
+				}()
+				expect(signal).to(beNil())
+				disposable?.dispose()
+				expect(signal).to(beNil())
+			}
+
+			it("should deallocate after erroring") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Cool

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+				var completed = false
+
+				signal?.observeCompleted { completed = true }
+
+				expect(completed).to(beFalsy())
+				expect(signal).toNot(beNil())
+
+				testScheduler.run()
+
+				expect(completed).to(beTruthy())
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Can you change all these to use the new syntax: `expect(completed) == true`, etc.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+//
+
+import Result
+import Nimble
+import Quick
+import ReactiveCocoa
+
+class SignalLifetimeSpec: QuickSpec {
+	override func spec() {
+		describe("init") {
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Maybe change this to `Signal.init`?

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+				expect(signal).to(beNil())
+				disposable?.dispose()
+				expect(signal).to(beNil())
+			}
+		}
+	}
+}
+
+private extension SignalType {
+
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Nit: unnecessary space

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+	func testTransform() -> Signal<Value, Error> {
+		return Signal { observer in
+			self.observe(observer.action)
+		}
+	}
+
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Nit: unnecessary space

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+			}
+		}
+	}
+}
+
+private extension SignalType {
+
+	func testTransform() -> Signal<Value, Error> {
+		return Signal { observer in
+			self.observe(observer.action)
```
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: this should be `return self.observe(observer.action)`

**Comment by NachoSoto** [comment]
Looks good! Just a few final comments on the tests.

File: ReactiveCocoa/Swift/Signal.swift (hunk 7 above)
Comment (Vadim-Yelagin) [inline ReactiveCocoa/Swift/Signal.swift]: No, something in ActionSpec.swift crashes. Apparently, disposable can be disposed after the signal is deallocated.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift (hunk 11 above)
Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: I don't know. There's `describe("init")` in SignalSpec and SignalProducerSpec too.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift (hunk 9 above)
Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: I've moved them from their original place in SignalSpec.swift as well.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Okay :)

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift (hunk 10 above)
Comment (Vadim-Yelagin) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: There's over 700 of `.to(beTruthy())` and `.to(beFalsy())` in the project. Since I only moved these cases, not written them, I think it would make more sense to either leave this as is, or replace all of those enties in a separate PR.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: I was thinking we can start changing them as we go over the tests, like here. But I understand your reasoning, feel free to leave as is!

**Comment by neilpa** [comment]
Looks good to me too

**Comment by NachoSoto** [comment]
:tada:! Thanks a lot @Vadim-Yelagin!
